src/storagetm/storagetm.c

```diff
@@ -0,0 +1,1058 @@
+        if (!p)
+                return log_oom();
+
+        *p = (NvmePort) {
+                .portnr = portnr,
+                .nvme_ports_fd = TAKE_FD(ports_fd),
+                .nvme_port_fd = TAKE_FD(port_fd),
+                .ip_family = ip_family,
+        };
+
+        *ret = TAKE_PTR(p);
+        return 0;
+}
+
+static int nvme_port_link_subsystem(NvmePort *port, NvmeSubsystem *subsys) {
+        assert(port);
+        assert(subsys);
+
+        _cleanup_free_ char *target = NULL, *linkname = NULL;
+        target = path_join("/sys/kernel/config/nvmet/subsystems", subsys->name);
+        if (!target)
+                return log_oom();
+
+        linkname = path_join("subsystems", subsys->name);
+        if (!linkname)
+                return log_oom();
+
+        if (symlinkat(target, port->nvme_port_fd, linkname) < 0)
+                return log_error_errno(errno, "Failed to link subsystem '%s' to port %" PRIu16 ": %m", subsys->name, port->portnr);
+
+        return 0;
+}
+
+static int nvme_port_unlink_subsystem(NvmePort *port, NvmeSubsystem *subsys) {
+        assert(port);
+        assert(subsys);
+
+        _cleanup_free_ char *linkname = NULL;
+        linkname = path_join("subsystems", subsys->name);
+        if (!linkname)
+                return log_oom();
+
+        if (unlinkat(port->nvme_port_fd, linkname, 0) < 0 && errno != ENOENT)
+                return log_error_errno(errno, "Failed to unlink subsystem '%s' to port %" PRIu16 ": %m", subsys->name, port->portnr);
+
+        return 0;
+}
+
+static int nvme_subsystem_report(NvmeSubsystem *subsystem, NvmePort *ipv4, NvmePort *ipv6) {
+        assert(subsystem);
+
+        _cleanup_free_ struct local_address *addresses = NULL;
+        int n_addresses;
+        n_addresses = local_addresses(NULL, 0, AF_UNSPEC, &addresses);
+        if (n_addresses < 0)
+                return log_error_errno(n_addresses, "Failed to determine local IP addresses: %m");
+
+        log_notice("NVMe-TCP: %s %s%s%s (%s)",
+                   special_glyph(SPECIAL_GLYPH_ARROW_RIGHT),
+                   emoji_enabled() ? special_glyph(SPECIAL_GLYPH_COMPUTER_DISK) : "", emoji_enabled() ? " " : "",
+                   subsystem->name, subsystem->device);
+
+        FOREACH_ARRAY(a, addresses, n_addresses) {
+                NvmePort *port = a->family == AF_INET ? ipv4 : ipv6;
+
+                if (!port)
+                        continue;
+
+                log_info("          %s Try for specific device: nvme connect -t tcp -n '%s' -a %s -s %" PRIu16,
+                         special_glyph(a >= addresses + (n_addresses - 1) ? SPECIAL_GLYPH_TREE_RIGHT : SPECIAL_GLYPH_TREE_BRANCH),
+                         subsystem->name,
+                         IN_ADDR_TO_STRING(a->family, &a->address),
+                         port->portnr);
+        }
+
+        return 0;
+}
+
+static int nvme_port_report(NvmePort *port) {
+        assert(port);
+
+        if (!port)
```

Comment:
This will never hit. Maybe, the assertion in the above is redundant??

Reply:
yeah. So, please drop the assertion. If IPv6 is off, the assertion will be triggered.